pkg/virt-exportserver/exportserver.go

```diff
@@ -0,0 +1,359 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package virtexportserver
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	goflag "flag"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"time"
+
+	flag "github.com/spf13/pflag"
+
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/service"
+)
+
+const (
+	authHeader = "x-kubevirt-export-token"
+)
+
+type TokenGetterFunc func() (string, error)
+
+type VolumeInfo struct {
+	Path       string
+	ArchiveURI string
+	DirURI     string
+	RawURI     string
+	RawGzURI   string
+}
+type ExportServerConfig struct {
+	Deadline time.Time
+
+	ListenAddr string
+
+	CertFile, KeyFile string
+
+	TokenFile string
+
+	Volumes []VolumeInfo
+
+	// unit testing helpers
+	ArchiveHandler func(string) http.Handler
+	DirHandler     func(string, string) http.Handler
+	FileHandler    func(string) http.Handler
+	GzipHandler    func(string) http.Handler
+
+	TokenGetter TokenGetterFunc
+}
+
+type execReader struct {
+	cmd    *exec.Cmd
+	stdout io.ReadCloser
+	stderr io.ReadCloser
+}
+
+type exportServer struct {
+	ExportServerConfig
+	handler http.Handler
+}
+
+func (er *execReader) Read(p []byte) (int, error) {
+	n, err := er.stdout.Read(p)
+	if err == io.EOF {
+		if err2 := er.cmd.Wait(); err2 != nil {
+			errBytes, _ := ioutil.ReadAll(er.stderr)
+			log.Log.Reason(err2).Errorf("Subprocess did not execute successfully, result is: %q\n%s", er.cmd.ProcessState.ExitCode(), string(errBytes))
+			return n, err2
+		}
+	}
+	return n, err
+}
+
+func (er *execReader) Close() error {
+	return er.stdout.Close()
+}
+
+func (s *exportServer) initHandler() {
+	mux := http.NewServeMux()
+	for _, vi := range s.Volumes {
+		for path, handler := range s.getHandlerMap(vi) {
+			log.Log.Infof("Handling path %s\n", path)
+			mux.Handle(path, tokenChecker(s.TokenGetter, handler))
+		}
+	}
+
+	s.handler = mux
+}
+
+func (s *exportServer) getHandlerMap(vi VolumeInfo) map[string]http.Handler {
+	var result = make(map[string]http.Handler)
+
+	if vi.ArchiveURI != "" {
+		result[vi.ArchiveURI] = s.ArchiveHandler(vi.Path)
+	}
+
+	if vi.DirURI != "" {
+		result[vi.DirURI] = s.DirHandler(vi.DirURI, vi.Path)
+	}
+
+	if vi.RawURI != "" {
+		fi, err := os.Stat(vi.Path)
+		if err != nil {
+			log.Log.Reason(err).Errorf("error statting %s", vi.Path)
+		} else {
+			if fi.IsDir() {
+				result[vi.RawURI] = s.FileHandler(path.Join(vi.Path, "disk.img"))
+			} else {
+				result[vi.RawURI] = s.FileHandler(vi.Path)
+			}
+		}
+	}
+
+	if vi.RawGzURI != "" {
+		fi, err := os.Stat(vi.Path)
+		if err != nil {
+			log.Log.Reason(err).Errorf("error statting %s", vi.Path)
+		} else {
+			if fi.IsDir() {
+				result[vi.RawGzURI] = s.GzipHandler(path.Join(vi.Path, "disk.img"))
+			} else {
+				result[vi.RawGzURI] = s.GzipHandler(vi.Path)
+			}
+		}
+	}
```

Comment:
Can we evaluate the if statment only once. Something like:
```go
        fi, err := os.Stat(vi.path)
        if err != nil {
                log.Log.Reason(err).Errorf("error statting %s", vi.path)
                return result
        }
        path := vi.path 
        if fi.IsDir() {
                path = gzipHandler(path.Join(vi.path, "disk.img"))
        }

        if vi.rawURI != "" {
                result[vi.rawURI] = fileHandler(path)
        }
        if vi.rawGzURI != "" {
                result[vi.rawGzURI] = gzipHandler(path)
        }
```

Reply:
yeah this code block can definitely be cleaned up